pkg/monitor/intervalcreation/podlogs.go

```diff
@@ -0,0 +1,145 @@
+			if interval.From.Before(beginning) || interval.From.After(end) ||
+				interval.To.Before(beginning) || interval.To.After(end) {
+				return nil, nil
+			}
+			return interval, err
+		}
+	}
+	return nil, nil
+}
+
+type etcdLogLine struct {
+	Level     string    `json:"level"`
+	Timestamp time.Time `json:"ts"`
+	Msg       string    `json:"msg"`
+}
+
+// etcdLogParser handles etcd logs which are already nicely json formatted such as:
+//
+// {"level":"info","ts":"2023-03-03T18:09:01.471Z","caller":"mvcc/index.go:214","msg":"compact tree index","revision":738215}
+func etcdLogParser(locator, line string, level monitorapi.EventLevel, logger logrus.FieldLogger) (*monitorapi.EventInterval, error) {
+	parsedLine := etcdLogLine{}
+	err := json.Unmarshal([]byte(line), &parsedLine)
+	if err != nil {
+		logger.WithError(err).Errorf("error parsing matched log line: %s", line)
+		return nil, err
+	}
+	return &monitorapi.EventInterval{
+		Condition: monitorapi.Condition{
+			Level:   level,
+			Locator: locator,
+			Message: parsedLine.Msg,
+		},
+		From: parsedLine.Timestamp,
+		To:   parsedLine.Timestamp.Add(1 * time.Second),
+	}, nil
+}
+
+func buildLogGatherers() []PodLogIntervalGenerator {
+	return []PodLogIntervalGenerator{
+		{
+			namespace: "openshift-etcd",
+			selector:  "app=etcd",
+			container: "etcd",
+			subStrings: []SubStringLevel{
+				{"slow fdatasync", monitorapi.Warning},
+			},
+			lineParser: etcdLogParser,
+		},
+	}
+}
+
+// IntervalsFromPodLogs fetches pod logs for a hardcoded set of namespace, label selector, and container. Each line is
+// then checked for a match of certain substrings and if found, passed to a function to parse the line to a
+// monitorapi.EventInterval, which will then be included in the main list of intervals.
+// Beginning and end times are specified so we only build intervals for the phase of testing we're interested in.
+// A single cluster in an upgrade job will have separate intervals for the upgrade phase and the conformance testing
+// phase.
+func IntervalsFromPodLogs(kubeClient kubernetes.Interface, beginning, end time.Time) (monitorapi.Intervals, error) {
+
+	intervals := monitorapi.Intervals{}
+	gatherers := buildLogGatherers()
+
+	for _, g := range gatherers {
+		podClient := kubeClient.CoreV1().Pods(g.namespace)
+		selector, _ := labels.Parse(g.selector)
+		pods, err := podClient.List(context.Background(), metav1.ListOptions{LabelSelector: selector.String()})
+		if err != nil {
+			logrus.WithError(err).Errorf("unable to list pods in %s namespace", g.namespace)
+			return nil, err
+		}
+		for _, pod := range pods.Items {
+			logger := logrus.WithField("pod", pod.Name)
+			logger.Infof("fetching logs between %s and %s", beginning.Format(time.RFC3339), end.Format(time.RFC3339))
+			reader, err := podClient.GetLogs(pod.Name, &kapiv1.PodLogOptions{Container: g.container}).Stream(context.Background())
+			if err != nil {
+				logger.WithError(err).Error("error reading pod logs")
```

Comment:
@dgoodwin coming back to this, I think it might be problematic only logging the error instead of interrupting the function, because otherwise if we error on getting some pod's logs we can get a nil reference

An example that we just encountered today: https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/periodic-ci-openshift-release-master-nightly-4.14-e2e-metal-ovn-assisted/1673609188414066688#1:build-log.txt%3A140

(of course no blaming here, it is extremely rare to have such a case and also not that easy to spot when reviewing golang code :sweat_smile: )